src/main/scala/viper/gobra/frontend/info/base/BuiltInMemberTag.scala

```diff
@@ -0,0 +1,339 @@
+    override def identifier: String = "Redeem"
+    override def name: String = "RedeemChannelMethodTag"
+    override def isPure: Boolean = false
+  }
+
+
+  /** Built-in MPredicate Tags */
+
+  case object IsChannelMPredTag extends BuiltInMPredicateTag {
+    override def identifier: String = "IsChannel"
+    override def name: String = "IsChannelMPredTag"
+  }
+
+  case object SendChannelMPredTag extends BuiltInMPredicateTag {
+    override def identifier: String = "SendChannel"
+    override def name: String = "SendChannelMPredTag"
+  }
+
+  case object RecvChannelMPredTag extends BuiltInMPredicateTag {
+    override def identifier: String = "RecvChannel"
+    override def name: String = "RecvChannelMPredTag"
+  }
+
+  case object ClosedMPredTag extends BuiltInMPredicateTag {
+    override def identifier: String = "Closed"
+    override def name: String = "ClosedMPredTag"
+  }
+
+  case object ClosureDebtMPredTag extends BuiltInMPredicateTag {
+    override def identifier: String = "ClosureDebt"
+    override def name: String = "ClosureDebtMPredTag"
+  }
+
+  case object TokenMPredTag extends BuiltInMPredicateTag {
+    override def identifier: String = "Token"
+    override def name: String = "TokenMPredTag"
+  }
+
+
+  /**
+    * Returns a vector of tags belonging to built-in members that should be considered during name resolution
+    */
+  def builtInMembers(): Vector[BuiltInMemberTag] = Vector(
+    // functions
+    CloseFunctionTag,
+    // fpredicates
+    PredTrueFPredTag,
+    // methods
+    SendGivenPermMethodTag,
+    SendGotPermMethodTag,
+    RecvGivenPermMethodTag,
+    RecvGotPermMethodTag,
+    InitChannelMethodTag,
+    CreateDebtChannelMethodTag,
+    RedeemChannelMethodTag,
+    // mpredicates
+    IsChannelMPredTag,
+    SendChannelMPredTag,
+    RecvChannelMPredTag,
+    ClosedMPredTag,
+    ClosureDebtMPredTag,
+    TokenMPredTag
+  )
+
+  def types(tag: BuiltInAuxTypeTag)(config: Config): AuxType = tag match {
+    // functions
+    case CloseFunctionTag => AuxType(
+      {
+        case (_, Vector(c: ChannelT, IntT(UnboundedInteger), IntT(UnboundedInteger)/* PermissionT */, PredT(Vector()))) if sendAndBiDirections.contains(c.mod) => noMessages
+        case (n, ts) => error(n, s"type error: close expects parameters of bidirectional or sending channel, int, int, and pred() types but got ${ts.mkString(", ")}")
+      },
+      {
+        case ts@Vector(c: ChannelT, IntT(UnboundedInteger), IntT(UnboundedInteger)/* PermissionT */, PredT(Vector())) if sendAndBiDirections.contains(c.mod) => FunctionT(ts, VoidType)
+      })
+    // fpredicates
+    case PredTrueFPredTag => AuxType(
+      {
+        case (_, _) => noMessages // it is well-defined for arbitrary arguments
+      },
+      {
+        case args => FunctionT(args, AssertionT)
+      })
+    case _ => unknownTagAuxType(tag)
+  }
+
+  def types(tag: BuiltInSingleAuxTypeTag)(config: Config): SingleAuxType = tag match {
```

Comment:
Having to extend monolithic methods (`types(...)`, `argGhostTyping(...)`, and `returnGhostTyping(...)`) when we add new tags seems a bit error-prone to me, despite the fact that this is well-documented in the beginning of the file. Is there any reason for these methods to not be defined in the traits `BuiltInAuxTypeTag` and `BuiltInSingleAuxTypeTag`? In such a scenario, we could modify those traits to look like
```scala
sealed trait BuiltInAuxTypeTag extends BuiltInMemberTag {
    def typ(config: Config): AuxType
    def argGhostTyping(args: Vector[Type])(config: Config): GhostType
    def returnGhostTyping(args: Vector[Type])(config: Config): GhostType
}
```
and similarly for `BuiltInAuxTypeTag`.

In that case, we would override the methods every time we define a new Tag and the compiler would enforce that we implement those methods.

Reply:
Great idea! If you go with `GhostBuiltInMemberTag` and it saves you from implementing `argGhostTyping` and `returnGhostTyping` since it can be derived (this corresponds to the previous "catch-all" case).